+adi/+AD400x/Base.m

```diff
@@ -0,0 +1,113 @@
+    % kSPS and 2 MSPS
+    %
+    % AD4000/AD4004/AD4008 are single-channel, 16-bit unsigned ADCs
+    % AD4001/AD4005 are single-channel, 16-bit signed ADCs
+    % AD4002/AD4006/AD4010 are single-channel, 18-bit unsigned ADCs
+    % AD4003/AD4007/AD4011 are single-channel, 18-bit signed ADCs
+
+    properties (Nontunable)
+        % SamplesPerFrame Samples Per Frame
+        %   Number of samples per frame, specified as an even positive
+        %   integer.
+        SamplesPerFrame = 4096
+
+	% SampleRate Sample Rate
+	%   Baseband sampling rate in Hz, specified as a scalar
+	%   in samples per second.
+	SampleRate = '500000'
+    end
+
+    properties (Dependent)
+        % Voltage
+        %   ADC Voltage in mV
+        Voltage
+
+        % VoltageScale Voltage Scale
+        %   ADC Voltage scale.
+        VoltageScale
+
+        % VoltageOffset Voltage Offset
+        %   ADC Voltage offset.
+        VoltageOffset
+    end
+
+    % Channel names
+    properties (Nontunable, Hidden, Constant)
+        channel_names = {'voltage0'}
+    end
+
+    % isOutput
+    properties (Hidden, Nontunable, Access = protected)
+        isOutput = false
+    end
+
+    properties (Nontunable, Hidden)
+        Timeout = Inf
+        kernelBuffersCount = 1
+        dataTypeStr
+        phyDevName
+        devName
+    end
+
+    properties (Nontunable, Hidden, Constant)
+        Type = 'Rx'
+    end
+
+    properties (Hidden, Constant)
+        ComplexData = false
+    end
+
+    methods
+        %% Constructor
+        function obj = Base(phydev, dev, dtype, varargin)
+            obj = obj@matlabshared.libiio.base(varargin{:});
+	    obj.phyDevName = phydev;
+	    obj.devName = dev;
+	    obj.dataTypeStr = dtype;
+        end
+
+	%% Set SampleRate
+	function set.SampleRate(obj, value)
+	    % Set device sampling rate
+	    obj.SampleRate = value;
+	    if obj.ConnectedToDevice
+	        obj.setDeviceAttributeRAW('sampling_frequency', num2str(value));
+	    end
+	end
+
+        %% Check Voltage
+        function rValue = get.Voltage(obj)
+            if obj.ConnectedToDevice
+                rValue = obj.getAttributeRAW('voltage0', 'raw', obj.isOutput);
+            else
+                rValue = NaN;
+            end
+        end
+
+        %% Check Voltage Scale
+        function rValue = get.VoltageScale(obj)
+            if obj.ConnectedToDevice
+                rValue = obj.getAttributeDouble('voltage0', 'scale', obj.isOutput);
+            else
+                rValue = NaN;
+            end
+        end
+
+        %% Check Voltage Offset
+        function rValue = get.VoltageOffset(obj)
+            if obj.ConnectedToDevice
+                rValue = obj.getAttributeDouble('voltage0', 'offset', obj.isOutput);
+            else
+                rValue = NaN;
+            end
+        end
+    end
+
+    %% API Functions
+    methods (Hidden, Access = protected)
+        function setupInit(~)
```

Comment:
Any attribute reads/writes performed before calling the system object won't be applied in the hardware. 

For the sampling rate attribute, should we cache the value in the property setter (if `ConnectedToDevice` is false), and invoke the setDeviceAttributeRAW method here in setupInit to apply it in the HW before the data capture?

Reply:
Hi, I've updated the commit to set the sampling frequency in this way in setupInit (based on what's happening in AD463x). Thanks for suggesting.

Reply:
Thanks, the function looks good now